#### Import Libraries

In [1]:
# Data manipulation
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

#### Loading the Dataset

In [2]:
df = pd.read_csv("../data/raw/online_shoppers_intention.csv")

#### Displaying the first 10 rows of the dataset to get a quick look at the data

In [4]:
df.head(10)

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
0,0,0.0,0,0.0,1,0.000000,0.200000,0.200000,0.0,0.0,Feb,1,1,1,1,Returning_Visitor,False,False
1,0,0.0,0,0.0,2,64.000000,0.000000,0.100000,0.0,0.0,Feb,2,2,1,2,Returning_Visitor,False,False
2,0,0.0,0,0.0,1,0.000000,0.200000,0.200000,0.0,0.0,Feb,4,1,9,3,Returning_Visitor,False,False
3,0,0.0,0,0.0,2,2.666667,0.050000,0.140000,0.0,0.0,Feb,3,2,2,4,Returning_Visitor,False,False
4,0,0.0,0,0.0,10,627.500000,0.020000,0.050000,0.0,0.0,Feb,3,3,1,4,Returning_Visitor,True,False
5,0,0.0,0,0.0,19,154.216667,0.015789,0.024561,0.0,0.0,Feb,2,2,1,3,Returning_Visitor,False,False
6,0,0.0,0,0.0,1,0.000000,0.200000,0.200000,0.0,0.4,Feb,2,4,3,3,Returning_Visitor,False,False
7,1,0.0,0,0.0,0,0.000000,0.200000,0.200000,0.0,0.0,Feb,1,2,1,5,Returning_Visitor,True,False
8,0,0.0,0,0.0,2,37.000000,0.000000,0.100000,0.0,0.8,Feb,2,2,2,3,Returning_Visitor,False,False
9,0,0.0,0,0.0,3,738.000000,0.000000,0.022222,0.0,0.4,Feb,2,4,1,2,Returning_Visitor,False,False


#### Displaying the total number of rows and columns in the dataset

In [5]:
rows, columns = df.shape
print(f'The dataset contains {rows} rows and {columns} columns.')

The dataset contains 12330 rows and 18 columns.


#### Displaying column names

In [6]:
df.columns

Index(['Administrative', 'Administrative_Duration', 'Informational',
       'Informational_Duration', 'ProductRelated', 'ProductRelated_Duration',
       'BounceRates', 'ExitRates', 'PageValues', 'SpecialDay', 'Month',
       'OperatingSystems', 'Browser', 'Region', 'TrafficType', 'VisitorType',
       'Weekend', 'Revenue'],
      dtype='str')

#### Based on the dataset documentation, the meaning of each column in the Online Shoppers Purchasing Intention dataset is as follows:

* **Administrative** - Number of administrative pages visited by the visitor during the session.
* **Administrative_Duration** - Total time spent by the visitor on administrative pages during the session.
* **Informational** - Number of informational pages visited by the visitor during the session.
* **Informational_Duration** - Total time spent by the visitor on informational pages during the session.
* **ProductRelated** - Number of product-related pages visited by the visitor during the session.
* **ProductRelated_Duration** - Total time spent by the visitor on product-related pages during the session.
* **BounceRates** - Average bounce rate of the pages visited during the session.
* **ExitRates** - Average exit rate of the pages visited during the session.
* **PageValues** - Average value of pages visited by the visitor before completing a transaction.
* **SpecialDay** - Closeness of the visit to a special day or shopping occasion, such as Valentine’s Day or Mother’s Day.
* **Month** - Month in which the shopping session occurred.
* **OperatingSystems** - Operating system used by the visitor during the session.
* **Browser** - Browser used by the visitor during the session.
* **Region** - Geographic region associated with the visitor.
* **TrafficType** - Type of traffic source through which the visitor reached the website.
* **VisitorType** - Type of visitor, such as returning visitor or new visitor.
* **Weekend** - Indicates whether the shopping session occurred on a weekend.
* **Revenue** - Indicates whether the shopping session resulted in a purchase, serving as the primary target variable for conversion classification.


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  str    
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType              12330 no

#### Displaying descriptive statistics for numerical variables

In [10]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Administrative,12330.0,2.315166,3.321784,0.0,0.000000,1.000000,4.000000,27.000000
Administrative_Duration,12330.0,80.818611,176.779107,0.0,0.000000,7.500000,93.256250,3398.750000
Informational,12330.0,0.503569,1.270156,0.0,0.000000,0.000000,0.000000,24.000000
Informational_Duration,12330.0,34.472398,140.749294,0.0,0.000000,0.000000,0.000000,2549.375000
ProductRelated,12330.0,31.731468,44.475503,0.0,7.000000,18.000000,38.000000,705.000000
ProductRelated_Duration,12330.0,1194.746220,1913.669288,0.0,184.137500,598.936905,1464.157214,63973.522230
BounceRates,12330.0,0.022191,0.048488,0.0,0.000000,0.003112,0.016813,0.200000
ExitRates,12330.0,0.043073,0.048597,0.0,0.014286,0.025156,0.050000,0.200000
PageValues,12330.0,5.889258,18.568437,0.0,0.000000,0.000000,0.000000,361.763742
SpecialDay,12330.0,0.061427,0.198917,0.0,0.000000,0.000000,0.000000,1.000000


#### Displaying columns along with their data types

In [11]:
print(df.dtypes)

Administrative               int64
Administrative_Duration    float64
Informational                int64
Informational_Duration     float64
ProductRelated               int64
ProductRelated_Duration    float64
BounceRates                float64
ExitRates                  float64
PageValues                 float64
SpecialDay                 float64
Month                          str
OperatingSystems             int64
Browser                      int64
Region                       int64
TrafficType                  int64
VisitorType                    str
Weekend                       bool
Revenue                       bool
dtype: object


#### Observations on Data Types

- The dataset contains a combination of numerical, categorical, and binary variables representing different aspects of online shopping sessions.
- Numerical variables are stored using int64 and float64 data types, while Month and VisitorType are stored as string (str) variables and Weekend and Revenue are stored as boolean (bool) variables.
- The current data types are generally appropriate for analysis, with categorical variables requiring suitable encoding during preprocessing for machine learning model development.

#### Checking unique values for each column

In [12]:
for col in df.columns:
    unique_vals = df[col].unique()

    print(f'Column Name : {col}')
    print(f'Number of Unique Values : {len(unique_vals)}')
    print(f'Unique Values : {unique_vals}')
    print('-' * 50)

Column Name : Administrative
Number of Unique Values : 27
Unique Values : [ 0  1  2  4 12  3 10  6  5  9  8 16 13 11  7 18 14 17 19 15 24 22 21 20
 23 27 26]
--------------------------------------------------
Column Name : Administrative_Duration
Number of Unique Values : 3335
Unique Values : [  0.         53.         64.6       ... 167.9107143 305.125
 150.3571429]
--------------------------------------------------
Column Name : Informational
Number of Unique Values : 17
Unique Values : [ 0  1  2  4 16  5  3 14  6 12  7  9 10  8 11 24 13]
--------------------------------------------------
Column Name : Informational_Duration
Number of Unique Values : 1258
Unique Values : [  0.   120.    16.   ... 547.75 368.25 211.25]
--------------------------------------------------
Column Name : ProductRelated
Number of Unique Values : 311
Unique Values : [  1   2  10  19   0   3  16   7   6  23  13  20   8   5  32   4  45  14
  52   9  46  15  22  11  12  36  42  27  90  18  38  17 128  25  30  21

#### Checking the target variable distribution

In [13]:
print("Revenue Value Counts:")
print(df['Revenue'].value_counts())

print("\nRevenue Percentage:")
print(df['Revenue'].value_counts(normalize=True) * 100)

Revenue Value Counts:
Revenue
False    10422
True      1908
Name: count, dtype: int64

Revenue Percentage:
Revenue
False    84.525547
True     15.474453
Name: proportion, dtype: float64


#### Checking missing values per column, including empty strings

In [14]:
# Count of NaN / None values per column
nan_count = df.isnull().sum()

# Count of empty strings per column
empty_count = (df == '').sum()

# Total missing values
total_missing = nan_count + empty_count

# Percentage of missing values
missing_percentage = (total_missing / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': total_missing,
    'Percentage (%)': missing_percentage
}, index=df.columns).sort_values(
    by='Percentage (%)',
    ascending=False
)

print(missing_df)

                         Missing Values  Percentage (%)
Administrative                        0             0.0
Administrative_Duration               0             0.0
Informational                         0             0.0
Informational_Duration                0             0.0
ProductRelated                        0             0.0
ProductRelated_Duration               0             0.0
BounceRates                           0             0.0
ExitRates                             0             0.0
PageValues                            0             0.0
SpecialDay                            0             0.0
Month                                 0             0.0
OperatingSystems                      0             0.0
Browser                               0             0.0
Region                                0             0.0
TrafficType                           0             0.0
VisitorType                           0             0.0
Weekend                               0         

#### Checking for duplicated rows

In [15]:
df.duplicated().sum()

np.int64(125)

#### Numerical Data Validation

(For this project, the continuous behavioral variables are especially important because they describe page activity, time spent, bounce behavior, exit behavior, and page value.)

In [16]:
numerical_columns = [
    'Administrative',
    'Administrative_Duration',
    'Informational',
    'Informational_Duration',
    'ProductRelated',
    'ProductRelated_Duration',
    'BounceRates',
    'ExitRates',
    'PageValues',
    'SpecialDay'
]

df[numerical_columns].describe()

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay
count,12330.000000,12330.000000,12330.000000,12330.000000,12330.000000,12330.000000,12330.000000,12330.000000,12330.000000,12330.000000
mean,2.315166,80.818611,0.503569,34.472398,31.731468,1194.746220,0.022191,0.043073,5.889258,0.061427
std,3.321784,176.779107,1.270156,140.749294,44.475503,1913.669288,0.048488,0.048597,18.568437,0.198917
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000,0.000000,7.000000,184.137500,0.000000,0.014286,0.000000,0.000000
50%,1.000000,7.500000,0.000000,0.000000,18.000000,598.936905,0.003112,0.025156,0.000000,0.000000
75%,4.000000,93.256250,0.000000,0.000000,38.000000,1464.157214,0.016813,0.050000,0.000000,0.000000
max,27.000000,3398.750000,24.000000,2549.375000,705.000000,63973.522230,0.200000,0.200000,361.763742,1.000000


#### Checking for negative values

(Most of these variables represent page counts, durations, rates, or values, so negative values would be invalid.)

In [17]:
(df[numerical_columns] < 0).sum()

Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
dtype: int64

#### Checking the range of behavioral rate variables

(Because BounceRates and ExitRates represent rates, we should explicitly verify whether their values fall within the expected 0–1 range.)

In [19]:
rate_columns = [
    'BounceRates',
    'ExitRates'
]

for col in rate_columns:
    print(f'{col}')
    print(f'Minimum: {df[col].min()}')
    print(f'Maximum: {df[col].max()}')
    print('-' * 40)

BounceRates
Minimum: 0.0
Maximum: 0.2
----------------------------------------
ExitRates
Minimum: 0.0
Maximum: 0.2
----------------------------------------


#### Checking categorical variables

In [20]:
categorical_columns = [
    'Month',
    'OperatingSystems',
    'Browser',
    'Region',
    'TrafficType',
    'VisitorType',
    'Weekend',
    'Revenue'
]

for col in categorical_columns:
    print(f'\n{col}')
    print(df[col].value_counts(dropna=False))


Month
Month
May     3364
Nov     2998
Mar     1907
Dec     1727
Oct      549
Sep      448
Aug      433
Jul      432
June     288
Feb      184
Name: count, dtype: int64

OperatingSystems
OperatingSystems
2    6601
1    2585
3    2555
4     478
8      79
6      19
7       7
5       6
Name: count, dtype: int64

Browser
Browser
2     7961
1     2462
4      736
5      467
6      174
10     163
8      135
3      105
13      61
7       49
12      10
11       6
9        1
Name: count, dtype: int64

Region
Region
1    4780
3    2403
4    1182
2    1136
6     805
7     761
9     511
8     434
5     318
Name: count, dtype: int64

TrafficType
TrafficType
2     3913
1     2451
3     2052
4     1069
13     738
10     450
6      444
8      343
5      260
11     247
20     198
9       42
7       40
15      38
19      17
14      13
18      10
16       3
12       1
17       1
Name: count, dtype: int64

VisitorType
VisitorType
Returning_Visitor    10551
New_Visitor           1694
Other                  

#### Validating binary variables

(Weekend and Revenue are expected to represent binary outcomes.)

In [21]:
print("Weekend unique values:")
print(df['Weekend'].unique())

print("\nRevenue unique values:")
print(df['Revenue'].unique())

Weekend unique values:
[False  True]

Revenue unique values:
[False  True]


#### Validating the target variable

In [22]:
print("Target Variable: Revenue")
print(f"Number of Classes: {df['Revenue'].nunique()}")

print("\nClass Distribution:")
print(df['Revenue'].value_counts())

print("\nClass Distribution (%):")
print((df['Revenue'].value_counts(normalize=True) * 100).round(2))

Target Variable: Revenue
Number of Classes: 2

Class Distribution:
Revenue
False    10422
True      1908
Name: count, dtype: int64

Class Distribution (%):
Revenue
False    84.53
True     15.47
Name: proportion, dtype: float64


#### Initial Data Quality Checks

In [23]:
print("Dataset Shape:", df.shape)
print("Missing Values:", df.isnull().sum().sum())
print("Duplicate Rows:", df.duplicated().sum())
print("Number of Features:", df.shape[1] - 1)
print("Target Variable:", 'Revenue')
print("Target Classes:", df['Revenue'].nunique())

Dataset Shape: (12330, 18)
Missing Values: 0
Duplicate Rows: 125
Number of Features: 17
Target Variable: Revenue
Target Classes: 2


#### Observations on Initial Data Quality

- The dataset contains 12,330 session records and 18 columns, including 17 features and the Revenue target variable for analyzing e-commerce purchase conversion.
- The dataset contains no missing values, ensuring complete information across all session-level features, while 125 duplicate rows require investigation and appropriate handling during data preparation.
- The Revenue variable contains 2 classes, establishing the problem as a binary classification task, with class imbalance requiring consideration during model development and evaluation.
